File: php/test/EntityTemplateEntityTest.php
<?php
declare(strict_types=1);

// EntityTemplate entity test

require_once __DIR__ . '/../shortcut_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class EntityTemplateEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = ShortcutSDK::test(null, null);
        $ent = $testsdk->EntityTemplate(null);
        $this->assertNotNull($ent);
    }

    // Feature #4: the entity stream(action, ...) method runs the op pipeline
    // and yields result items. With the streaming feature active it yields the
    // feature's incremental output; otherwise it falls back to the materialised
    // list so stream always yields.
    public function test_stream(): void
    {
        $seed = [
            "entity" => [
                "entity_template" => [
                    "s1" => ["id" => "s1"],
                    "s2" => ["id" => "s2"],
                    "s3" => ["id" => "s3"],
                ],
            ],
        ];

        // Fallback: streaming inactive -> yields the materialised list items.
        $base = ShortcutSDK::test($seed, null);
        $seen = iterator_to_array($base->EntityTemplate(null)->stream("list", null, null), false);
        $this->assertCount(3, $seen);

        // Inbound: streaming active -> yields each item from the feature.
        $cfg = ShortcutConfig::make_config();
        if (isset($cfg["feature"]) && is_array($cfg["feature"]) && isset($cfg["feature"]["streaming"])) {
            $sdk = ShortcutSDK::test($seed, ["feature" => ["streaming" => ["active" => true]]]);
            $got = [];
            foreach ($sdk->EntityTemplate(null)->stream("list", null, null) as $item) {
                if (is_array($item) && array_is_list($item)) {
                    foreach ($item as $sub) {
                        $got[] = $sub;
                    }
                } else {
                    $got[] = $item;
                }
            }
            $this->assertCount(3, $got);
        }
    }

    public function test_basic_flow(): void
    {
        $setup = entity_template_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["create", "list", "update", "load", "remove"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "entity_template." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set SHORTCUT_TEST_ENTITY_TEMPLATE_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // CREATE
        $entity_template_ref01_ent = $client->EntityTemplate(null);
        $entity_template_ref01_data = Helpers::to_map(Vs::getprop(
            Vs::getpath($setup["data"], "new.entity_template"), "entity_template_ref01"));

        $entity_template_ref01_data_result = $entity_template_ref01_ent->create($entity_template_ref01_data, null);
        $entity_template_ref01_data = Helpers::to_map($entity_template_ref01_data_result);
        $this->assertNotNull($entity_template_ref01_data);
        $this->assertNotNull($entity_template_ref01_data["id"]);

        // LIST
        $entity_template_ref01_match = [];

        $entity_template_ref01_list_result = $entity_template_ref01_ent->list($entity_template_ref01_match, null);
        $this->assertIsArray($entity_template_ref01_list_result);

        $found_item = sdk_select(
            Runner::entity_list_to_data($entity_template_ref01_list_result),
            ["id" => $entity_template_ref01_data["id"]]);
        $this->assertNotEmpty($found_item);

        // UPDATE
        $entity_template_ref01_data_up0_up = [
            "id" => $entity_template_ref01_data["id"],
        ];

        $entity_template_ref01_markdef_up0_name = "author_id";
        $entity_template_ref01_markdef_up0_value = "Mark01-entity_template_ref01_" . $setup["now"];
        $entity_template_ref01_data_up0_up[$entity_template_ref01_markdef_up0_name] = $entity_template_ref01_markdef_up0_value;

        $entity_template_ref01_resdata_up0_result = $entity_template_ref01_ent->update($entity_template_ref01_data_up0_up, null);
        $entity_template_ref01_resdata_up0 = Helpers::to_map($entity_template_ref01_resdata_up0_result);
        $this->assertNotNull($entity_template_ref01_resdata_up0);
        $this->assertEquals($entity_template_ref01_resdata_up0["id"], $entity_template_ref01_data_up0_up["id"]);
        $this->assertEquals($entity_template_ref01_resdata_up0[$entity_template_ref01_markdef_up0_name], $entity_template_ref01_markdef_up0_value);

        // LOAD
        $entity_template_ref01_match_dt0 = [
            "id" => $entity_template_ref01_data["id"],
        ];
        $entity_template_ref01_data_dt0_loaded = $entity_template_ref01_ent->load($entity_template_ref01_match_dt0, null);
        $entity_template_ref01_data_dt0_load_result = Helpers::to_map($entity_template_ref01_data_dt0_loaded);
        $this->assertNotNull($entity_template_ref01_data_dt0_load_result);
        $this->assertEquals($entity_template_ref01_data_dt0_load_result["id"], $entity_template_ref01_data["id"]);

        // REMOVE
        $entity_template_ref01_match_rm0 = [
            "id" => $entity_template_ref01_data["id"],
        ];
        $entity_template_ref01_ent->remove($entity_template_ref01_match_rm0, null);

        // LIST
        $entity_template_ref01_match_rt0 = [];

        $entity_template_ref01_list_rt0_result = $entity_template_ref01_ent->list($entity_template_ref01_match_rt0, null);
        $this->assertIsArray($entity_template_ref01_list_rt0_result);

        $not_found_item = sdk_select(
            Runner::entity_list_to_data($entity_template_ref01_list_rt0_result),
            ["id" => $entity_template_ref01_data["id"]]);
        $this->assertEmpty($not_found_item);

    }
}

function entity_template_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/entity_template/EntityTemplateTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = ShortcutSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["entity_template01", "entity_template02", "entity_template03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("SHORTCUT_TEST_ENTITY_TEMPLATE_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "SHORTCUT_TEST_ENTITY_TEMPLATE_ENTID" => $idmap,
        "SHORTCUT_TEST_LIVE" => "FALSE",
        "SHORTCUT_TEST_EXPLAIN" => "FALSE",
        "SHORTCUT_APIKEY" => "NONE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["SHORTCUT_TEST_ENTITY_TEMPLATE_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["SHORTCUT_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
                "apikey" => $env["SHORTCUT_APIKEY"],
            ],
            $extra ?? [],
        ]);
        $client = new ShortcutSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["SHORTCUT_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["SHORTCUT_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}

File: lua/test/entity_template_entity_test.lua
-- EntityTemplate entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("shortcut_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("EntityTemplateEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:EntityTemplate(nil)
    assert.is_not_nil(ent)
  end)

  -- Feature #4: the entity stream(action, ...) method runs the op pipeline and
  -- returns an iterator over result items. With the streaming feature active it
  -- yields the feature's incremental output; otherwise it falls back to the
  -- materialised list so stream always yields.
  it("should stream", function()
    local seed = {
      entity = {
        ["entity_template"] = {
          s1 = { id = "s1" },
          s2 = { id = "s2" },
          s3 = { id = "s3" },
        },
      },
    }

    -- Fallback: streaming inactive -> yields the materialised list items.
    local base = sdk.test(seed, nil)
    local seen = {}
    for item in base:EntityTemplate(nil):stream("list", nil, nil) do
      table.insert(seen, item)
    end
    assert.are.equal(3, #seen)

    -- Inbound: streaming active -> yields each item from the feature.
    local config = require("config")()
    if type(config.feature) == "table" and config.feature.streaming ~= nil then
      local streamsdk = sdk.test(seed, { feature = { streaming = { active = true } } })
      local got = {}
      for item in streamsdk:EntityTemplate(nil):stream("list", nil, nil) do
        if vs.islist(item) then
          for _, sub in ipairs(item) do
            table.insert(got, sub)
          end
        else
          table.insert(got, item)
        end
      end
      assert.are.equal(3, #got)
    end
  end)

  it("should run basic flow", function()
    local setup = entity_template_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"create", "list", "update", "load", "remove"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "entity_template." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set SHORTCUT_TEST_ENTITY_TEMPLATE_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- CREATE
    local entity_template_ref01_ent = client:EntityTemplate(nil)
    local entity_template_ref01_data = helpers.to_map(vs.getprop(
      vs.getpath(setup.data, "new.entity_template"), "entity_template_ref01"))

    local entity_template_ref01_data_result, err = entity_template_ref01_ent:create(entity_template_ref01_data, nil)
    assert.is_nil(err)
    entity_template_ref01_data = helpers.to_map(entity_template_ref01_data_result)
    assert.is_not_nil(entity_template_ref01_data)
    assert.is_not_nil(entity_template_ref01_data["id"])

    -- LIST
    local entity_template_ref01_match = {}

    local entity_template_ref01_list_result, err = entity_template_ref01_ent:list(entity_template_ref01_match, nil)
    assert.is_nil(err)
    assert.is_table(entity_template_ref01_list_result)

    local found_item = vs.select(
      runner.entity_list_to_data(entity_template_ref01_list_result),
      { id = entity_template_ref01_data["id"] })
    assert.is_false(vs.isempty(found_item))

    -- UPDATE
    local entity_template_ref01_data_up0_up = {
      id = entity_template_ref01_data["id"],
    }

    local entity_template_ref01_markdef_up0_name = "author_id"
    local entity_template_ref01_markdef_up0_value = "Mark01-entity_template_ref01_" .. tostring(setup.now)
    entity_template_ref01_data_up0_up[entity_template_ref01_markdef_up0_name] = entity_template_ref01_markdef_up0_value

    local entity_template_ref01_resdata_up0_result, err = entity_template_ref01_ent:update(entity_template_ref01_data_up0_up, nil)
    assert.is_nil(err)
    local entity_template_ref01_resdata_up0 = helpers.to_map(entity_template_ref01_resdata_up0_result)
    assert.is_not_nil(entity_template_ref01_resdata_up0)
    assert.are.equal(entity_template_ref01_resdata_up0["id"], entity_template_ref01_data_up0_up["id"])
    assert.are.equal(entity_template_ref01_resdata_up0[entity_template_ref01_markdef_up0_name], entity_template_ref01_markdef_up0_value)

    -- LOAD
    local entity_template_ref01_match_dt0 = {
      id = entity_template_ref01_data["id"],
    }
    local entity_template_ref01_data_dt0_loaded, err = entity_template_ref01_ent:load(entity_template_ref01_match_dt0, nil)
    assert.is_nil(err)
    local entity_template_ref01_data_dt0_load_result = helpers.to_map(entity_template_ref01_data_dt0_loaded)
    assert.is_not_nil(entity_template_ref01_data_dt0_load_result)
    assert.are.equal(entity_template_ref01_data_dt0_load_result["id"], entity_template_ref01_data["id"])

    -- REMOVE
    local entity_template_ref01_match_rm0 = {
      id = entity_template_ref01_data["id"],
    }
    local _, err = entity_template_ref01_ent:remove(entity_template_ref01_match_rm0, nil)
    assert.is_nil(err)

    -- LIST
    local entity_template_ref01_match_rt0 = {}

    local entity_template_ref01_list_rt0_result, err = entity_template_ref01_ent:list(entity_template_ref01_match_rt0, nil)
    assert.is_nil(err)
    assert.is_table(entity_template_ref01_list_rt0_result)

    local not_found_item = vs.select(
      runner.entity_list_to_data(entity_template_ref01_list_rt0_result),
      { id = entity_template_ref01_data["id"] })
    assert.is_true(vs.isempty(not_found_item))

  end)
end)

function entity_template_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/entity_template/EntityTemplateTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read entity_template test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "entity_template01", "entity_template02", "entity_template03" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("SHORTCUT_TEST_ENTITY_TEMPLATE_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["SHORTCUT_TEST_ENTITY_TEMPLATE_ENTID"] = idmap,
    ["SHORTCUT_TEST_LIVE"] = "FALSE",
    ["SHORTCUT_TEST_EXPLAIN"] = "FALSE",
    ["SHORTCUT_APIKEY"] = "NONE",
  })

  local idmap_resolved = helpers.to_map(
    env["SHORTCUT_TEST_ENTITY_TEMPLATE_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end

  if env["SHORTCUT_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      {
        apikey = env["SHORTCUT_APIKEY"],
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["SHORTCUT_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["SHORTCUT_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end
